app/src/lib/popup-manager.ts

```diff
@@ -98,11 +109,25 @@ export class PopupManager {
       return popupToAdd
     }
 
-    this.popupStack.push(popup)
+    this.insertBeforeErrorPopups(popup)
     this.checkStackLength()
     return popup
   }
 
+  /** Adds a non-Error type popup before any error popups. */
+  private insertBeforeErrorPopups(popup: Popup) {
+    const indexLastError = this.popupStack.findIndex(
+      p => p.type === PopupType.Error
+    )
+
+    if (indexLastError === -1) {
+      this.popupStack.push(popup)
+      return
+    }
+
+    this.popupStack.splice(indexLastError, 0, popup)
```

Comment:
Now that we expose the internal `popupStack` directly via `allPopups` we need to ensure that it doesn't mutate. Consumers receiving a ReadonlyArray have a reasonable expectation that it won't change underneath them which can lead to subtle bugs in a snapshot-state model like ours. We should make the private `popupStack` be a `ReadonlyArray` as well to enforce this within this class.

```suggestion
    if (this.popupStack.at(-1)?.type !== PopupType.Error) {
      this.popupStack = this.popupStack.concat(popup)
      return
    }

    const errorPopups = this.getPopupsOfType(PopupType.Error)
    const nonErrorPopups = this.popupStack.filter(
      p => p.type !== PopupType.Error
    )
    this.popupStack = [...nonErrorPopups, popup, ...errorPopups]
```